### DISCUSSION TITLE
START MIGRATION REWRITE / COMMIT MIGRATION REWRITE

### DISCUSSION THREAD

**Discussion by msullivan** [geldata/gel#4562] [Pre-RFC]
Edit: We are doing this, as described in this edited version

I am proposing a fairly low-surface area set of features for enabling history rewrite. I think this is approximately what we discussed?

`START MIGRATION REWRITE`:
Enters a migration history rewrite mode. This behaves like `START MIGRATION` in many ways. If there is not a current transaction, this begins an implicit one, like `START MIGRATION` (though we *could* drop that without it being a real problem).

At the start of the migration history rewrite mode, the current effective schema becomes *empty*. The client can then perform migrations to try to advance the schema to match the actual schema state. Importantly, `START MIGRATION` may be nested inside a `START MIGRATION REWRITE`. `CREATE MIGRATION` may also be explicitly executed.

`COMMIT MIGRATION REWRITE`:
Commit the migration history rewrite.

The effective schema in the history rewrite mode is compared against the actual schema. If they do not match, we produce an error.

If they do match, then all entries in schema::Migration are dropped and replaced with new ones representing the migrations performed in the rewrite mode, and we leave the rewrite mode.

No other actions are performed: no DDL or queries are executed.

`ABORT MIGRATION REWRITE`:
Abort the current rewrite. This isn't *really* needed (you could just use a savepoint explicitly), but it mirrors `ABORT MIGRATION`.

`START MIGRATION TO COMMITTED SCHEMA`:
Start a migration to the target schema of the of the rewrite. This isn't strictly necessary, since the client could provide that full schema

----

I tested the syntax in our parser, it's all fine.

There are possible features that are left out of this design:

* Starting the rewrite from a particular point in the migration history: we don't store historical schemas, and so the implementation of this would need to parse and apply all prior migrations anyway. This wouldn't be a real improvement over the CLI just doing that.

* An equivalent to `DESCRIBE CURRENT MIGRATION AS JSON`. The client can use `START MIGRATION TO COMMITTED SCHEMA` and then `DESCRIBE` (+ `POPULATE MIGRATION` even) to compare the current state to the target state, and then abort that migration.

**Comment by elprans** [comment]
> An equivalent to `DESCRIBE CURRENT MIGRATION AS JSON`. 

Maybe just repurpose `DESCRIBE SCHEMA`?

**Reply by tailhook** [reply]
Well, I don't understand what that sentence means. Do both of you mean `DESCRIBE SCHEMA OF REWRITE TARGET` ? Anyone can do that before starting or on parallel connection.

If it's about diff. Then yes, `START MIGRATION TO` is a way to diff in all modes. If we find any way to emit a diff besides a set of DDL statements, then yes, we can add a new statement.

**Reply by elprans** [reply]
No, I literally meant `DESCRIBE SCHEMA`.  Entering a rewrite context alters the behavior of all DDL statements, so, perhaps, referring to the "real" schema should be the special case, not the other way around?

**Reply by tailhook** [reply]
I still don't understand what you mean. `DESCRIBE SCHEMA` should definitely describe what's the schema in the current context, not the committed schema. As I've said in the comment above, committed schema doesn't need any special syntax.

And how is that related to `DESCRIBE CURRENT MIGRATION AS JSON` ?

**Comment by tailhook** [comment]
> If there is not a current transaction, this begins an implicit one, like START MIGRATION (though we could drop that without it being a real problem).

Just a note: I think the real consequence of this behavior is: once you've got a failed query, your transaction is failed and doesn't accept any queries until `ROLLBACK`. And this is communicated to the client via transaction status.

So we can't "drop that" without introducing extra state protocol or changing semantics (i.e. query failures being a non-issue server-side, not sure what are consequences of this).

**Comment by tailhook** [comment]
Okay so this solves my current use cases. Comparing it with the "abstract databases" concept (from user point of view):

Upsides:
1. You don't need an extra connection
2. Final start migration doesn't needs original schema sent

Downsides:
1. Abstract databases could be used on lightweight/headless EdgeDB server instances (I mean instances having no postgres). Use cases: language server (LSP), syntax check for the schema+migrations, and checking that migrations are compatible with multiple EdgeDB versions on CI, codegen.

So fine with me. But I feel that we will need to reinvent headless EdgeDB server anyways.

**Comment by tailhook** [comment]
One point to bikeshed, I'm not sure `TO REWRITE TARGET` reads well. Maybe:
* `START MIGRATION TO ORIGINAL SCHEMA`
* `START MIGRATION TO TARGET SCHEMA`
* `START MIGRATION TO FINAL SCHEMA`

Well, I kinda have tiny concerns of both of these. Maybe you can propose another option?

I don't like `TO REWRITE TARGET` because we're basically rewriting history, but we are migrating to some schema. Also it's quite implicit what target schema is.

**Reply by elprans** [reply]
I think we need to add support for `FROM` to allow generating migrations between arbitrary schemas.  Then, known schemas can be identified by special keywords, e.g:

- `CURRENT SCHEMA` -- current "real" schema or current rewritten schema (if inside `START MIGRATION REWRITE`)
- `COMMITTED SCHEMA` -- current "real" schema persisted in backend.

So, from within `START MIGRATION REWRITE` you can `START MIGRATION FROM COMMITTED SCHEMA TO CURRENT SCHEMA` to get a diff and generate a migration from real schema to rewritten schema.

**Reply by msullivan** [reply]
What's the use case for this? It doesn't sound like the CLI needs it for what we have been discussing?

It's definitely *doable*, it's just kind of weird because a `START MIGRATION FROM` can never be committed

**Reply by elprans** [reply]
> What's the use case for this? It doesn't sound like the CLI needs it for what we have been discussing?

I think we need to be consistent what "current schema" means in the context of `START MIGRATION REWRITE`.  IMO, it should be the schema you're working on, not the "committed" schema.  Hence, at the very least, the CLI needs to be able to compare the "rewritten" schema to "committed" schema, ideally without having to send and parse the entire SDL (which could be unavailable)?

>  it's just kind of weird because a `START MIGRATION FROM` can never be committed

We can only allow this form in `START MIGRATION REWRITE`

**Reply by msullivan** [reply]
> > What's the use case for this? It doesn't sound like the CLI needs it for what we have been discussing?
> 
> I think we need to be consistent what "current schema" means in the context of `START MIGRATION REWRITE`. IMO, it should be the schema you're working on, not the "committed" schema. Hence, at the very least, the CLI needs to be able to compare the "rewritten" schema to "committed" schema, ideally without having to send and parse the entire SDL (which could be unavailable)?
> 
Sure, but does it need to be able to do it *bidirectionally*? 

What I called `START MIGRATION TO REWRITE TARGET` already does compare the "rewritten" schema to the "committed" schema. I do think that `START MIGRATION TO COMMITTED SCHEMA` would be a better name for it.

> > it's just kind of weird because a `START MIGRATION FROM` can never be committed
> 
> We can only allow this form in `START MIGRATION REWRITE`

Sure, but any other migration inside a rewrite block you can still commit.

**Reply by tailhook** [reply]
I like `TO COMMITTED SCHEMA`

**Reply by tailhook** [reply]
Regarding `FROM`, off the top of my head this can be used to generate "downgrade migration" file (i.e. from schema in the filesystem to the comitted schema). But I'm not sure if this a good way to approach the problem of "down" migrations.
